### PR TITLE
always back `ty::Const` by an `Allocation`

### PR DESCRIPTION
**Pull request by oli-obk**
This reduces the number of conversion between `Allocation` and `Value` by caching both.

r? @RalfJung

### CODE REVIEW

File: src/librustc/mir/interpret/value.rs [hunk 1]
```diff
@@ -2,21 +2,24 @@ use std::fmt;
...
 pub struct RawConst<'tcx> {
-    // the value lives here, at offset 0, and that allocation definitely is a `AllocKind::Memory`
-    // (so you can use `AllocMap::unwrap_memory`).
+    /// the value lives here, at offset 0, and the allocation that it refers to is the one in the
+    /// `alloc` field
     pub alloc_id: AllocId,
     pub ty: Ty<'tcx>,
+    /// the allocation that would be returned by using
+    /// `tcx.alloc_map.lock().unwrap_memory(self.alloc_id)`
+    pub alloc: &'tcx Allocation,
```
Comment (RalfJung) [inline src/librustc/mir/interpret/value.rs]: Why is that needed? Is it just an optimization?

Please answer by pushing an extended comment. ;)

File: src/librustc/ty/sty.rs [hunk 2]
```diff
@@ -2065,7 +2066,7 @@ pub enum LazyConst<'tcx> {
 }
 
 #[cfg(target_arch = "x86_64")]
-static_assert!(LAZY_CONST_SIZE: ::std::mem::size_of::<LazyConst<'static>>() == 56);
+static_assert!(LAZY_CONST_SIZE: ::std::mem::size_of::<LazyConst<'static>>() == 80);
```
Comment (RalfJung) [inline src/librustc/ty/sty.rs]: And this isn't a problem? Let's do a perf run.

File: src/librustc/ty/sty.rs [hunk 3]
```diff
@@ -2086,15 +2087,66 @@ impl<'tcx> LazyConst<'tcx> {
...
+    /// literals)
+    pub val: ConstValue,
+
+    /// The actual backing storage of the constant and a pointer which can be resolved back to the
+    /// `allocation` field
+    ///
+    /// Can be `None` for trivial constants created from literals or directly. Is always `Some` for
+    /// aggregate constants or any named constant that you can actually end up taking a reference
+    /// to. This will get unwrapped in situations where we do know that it's a referencable
+    pub alloc: Option<(&'tcx Allocation, Pointer)>,
```
Comment (RalfJung) [inline src/librustc/ty/sty.rs]: Hm, I think I'd prefer if this would not have to be an `Option` -- but if that has a non-zero cost, and if this does not actually cause any problems in practice, then whatever.

File: src/librustc_mir/const_eval.rs [hunk 4]
```diff
@@ -96,22 +96,25 @@ pub fn op_to_const<'tcx>(
...
         Ok(Immediate::Scalar(x)) =>
-            ConstValue::Scalar(x.not_undef()?),
+            (ConstValue::Scalar(x.not_undef()?), None),
         Ok(Immediate::ScalarPair(a, b)) =>
-            ConstValue::Slice(a.not_undef()?, b.to_usize(ecx)?),
+            (ConstValue::Slice(a.not_undef()?, b.to_usize(ecx)?), None),
     };
-    Ok(ty::Const { val, ty: op.layout.ty })
+    Ok(ty::Const { val, ty: op.layout.ty, alloc })
 }
```
Comment (RalfJung) [inline src/librustc_mir/const_eval.rs]: Where is `op_to_const` even still needed? Isn't the purpose of this patch to get rid of such awful conversion functions?

Comment (oli-obk) [reply src/librustc_mir/const_eval.rs]: This is the groundwork for that. Getting rid of the conversion functions needs some more refactorings in the match code.

File: src/librustc/ty/sty.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc/ty/sty.rs]: It's actually somewhat problematic to not use an option, because we generate a bunch of simple constants like 1 and 0 in the match code. These constants never leak to the rest of the compiler afaik, but it simplifies the code a lot to use the same datastructure

File: src/librustc_mir/const_eval.rs [hunk 5]
```diff
@@ -536,9 +538,17 @@ fn validate_and_turn_into_const<'a, 'tcx>(
             )?;
         }
         // Now that we validated, turn this into a proper constant.
-        let def_id = cid.instance.def.def_id();
-        let normalize = tcx.is_static(def_id).is_none() && cid.promoted.is_none();
-        op_to_const(&ecx, op, normalize)
+        let val = match op.layout.abi {
+            layout::Abi::Scalar(..) => ConstValue::Scalar(ecx.read_immediate(op)?.to_scalar()?),
```
Comment (RalfJung) [inline src/librustc_mir/const_eval.rs]: This is the key point where we decide what can get "cached" in `val`, right? Might be worth adding a comment explaining what's going on.

File: src/librustc/ty/sty.rs (hunk 3 above)
Comment (RalfJung) [reply src/librustc/ty/sty.rs]: Okay, makes sense, I guess.

File: src/librustc_mir/const_eval.rs (hunk 4 above)
Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: Oh I see.

What I do not like is that this PR now duplicates the logic about what gets "cached" in `val`. Can you avoid that?

Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: But, actually, why can't `const_field` create an `MPlace` from the `alloc` stuff, and thus entirely avoid `lazy_const_to_op` and `op_to_const`?

Comment (oli-obk) [reply src/librustc_mir/const_eval.rs]: It can. I just really dislike doing many such refactorings at once.

I'll try to avoid duplicating the caching logic

Comment (oli-obk) [reply src/librustc_mir/const_eval.rs]: deduplicating did not succeed, I will try again once all the `Op` conversion stuff is simplified ot `MemPlace`

Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: > I just really dislike doing many such refactorings at once.

Well, that is a completely local change in `const_field`, is it not?  And it proves the viability of the refactoring, so it seems like a good sanity check.

File: src/librustc/ich/impls_ty.rs [hunk 6]
```diff
@@ -372,10 +375,30 @@ impl_stable_hash_for!(enum ::syntax::ast::Mutability {
...
+        &self,
+        hcx: &mut StableHashingContext<'a>,
+        hasher: &mut StableHasher<W>,
+    ) {
+        let ty::Const { ty, val, alloc } = self;
+        ty.hash_stable(hcx, hasher);
+        val.hash_stable(hcx, hasher);
+        // don't hash the memory for `Scalar` and `Slice`. There's nothing to be gained
+        // by it. All the relevant info is contained in the value.
+        if let mir::interpret::ConstValue::ByRef = val {
```
Comment (RalfJung) [inline src/librustc/ich/impls_ty.rs]: Uh... so, I know nothing about this stuff.^^ But, isn't it a problem that now we will equate the same constant, when computed multiple times and stored in different allocations? Like, pointer equality might matter for some of them.

File: src/librustc_mir/const_eval.rs (hunk 4 above)
Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: If this is *not* just a local change, I am fine doing this in separate PRs, but then I'd like to understand why that is the case. :)

**Comment by oli-obk** [comment]
@rust-timer build dc96b91d61feea271e00df81401bebae68770918

File: src/librustc/ich/impls_ty.rs (hunk 6 above)
Comment (oli-obk) [reply src/librustc/ich/impls_ty.rs]: Well.. the *value* would be the same. Once you have a `ty::Const`, there is no identity anymore. You can't take a reference to a `ty::Const`, so you can't turn it back into a reference.

File: src/librustc_mir/const_eval.rs (hunk 4 above)
Comment (oli-obk) [reply src/librustc_mir/const_eval.rs]: I just tried it out.... I can indeed avoid `op_to_const` completely, but `lazy_const_to_op` is still necessary for `const_discriminant`, because `read_discriminant` takes an `OpTy`.

Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: Well, that's a start. :)

Still, I am confused. I assume you have some kind of `const_to_mplace` for this? It can take a `ty::Const`, that wraps up a `Ty` and the allocation stuff so we can compute the layout, which is great. Once you have an `MPlaceTy`, you can use all the great projection functions, and you can call `into()` to get an `OpTy` for `read_discriminant`.

Comment (oli-obk) [reply src/librustc_mir/const_eval.rs]: While we can do that, it will mean that reading the discriminant from an `Option<&T>` will need to go through the memory instead of just scraping it from the value directly.

Oh... that happens for the field access, too. Well.. we can just bench it and see how it goes

Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: Hm, yes, that's true.

But I just realized that maybe we can have a `lazy_const_to_op` that's not quite as bad because it doesn't take a `ty`? The `Evaluated` case contains a `Ty`, after all, and the `Unevaluated` case should have all the information to get it. Or can we even exclude `Unevaluated` entirely because we only use this in pattern matching on already evaluated stuff? `const_to_op` (taking `ty::Const` seems fairly reasonable, *if* it does not take a `Ty` or `TyLayout`.

Comment (oli-obk) [reply src/librustc_mir/const_eval.rs]: There's still a use of `lazy_const_to_op` left in the const propagator, and I'm at a total loss how to get rid of that one. I'll implement your suggestion.

Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: Hm... looking at that, can't it use `ecx.eval_operand(Operand::Constant(Box::new(c)))`?

**Comment by RalfJung** [comment]
[max-rss regressed noticeably](https://perf.rust-lang.org/compare.html?start=f058741a679062202697144daebd05e1b0d6d747&end=dc96b91d61feea271e00df81401bebae68770918&stat=max-rss), mostly for tuple-stress.

File: src/librustc_mir/interpret/memory.rs [hunk 7]
```diff
@@ -647,15 +646,15 @@ where
             // does not permit code that would break this!
             if self.alloc_map.contains_key(&alloc) {
                 // Not yet interned, so proceed recursively
-                self.intern_static(alloc, mutability)?;
+                let _alloc = self.intern_static(alloc, mutability)?;
```
Comment (RalfJung) [inline src/librustc_mir/interpret/memory.rs]: Why bind this to an unused variable?

File: src/librustc_mir/const_eval.rs [hunk 8]
```diff
@@ -475,21 +434,45 @@ pub fn const_field<'a, 'tcx>(
...
         };
         // then project
-        let field = ecx.operand_field(down, field.index() as u64)?;
-        // and finally move back to the const world, always normalizing because
-        // this is not called for statics.
-        op_to_const(&ecx, field, true)
+        let field = ecx.mplace_field(down, field.index() as u64)?;
+        let val = match field.layout.abi {
+            layout::Abi::Scalar(..) => {
+                let scalar = ecx.try_read_immediate_from_mplace(field)?.unwrap().to_scalar()?;
```
Comment (RalfJung) [inline src/librustc_mir/const_eval.rs]: Couldn't you use `try_read_immediate(field.into())`? Then we could avoid exposing that internal function `try_read_immediate_from_mplace`.

File: src/librustc_mir/interpret/step.rs [hunk 9]
```diff
@@ -266,8 +266,8 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
             }
 
             Discriminant(ref place) => {
-                let place = self.eval_place(place)?;
-                let discr_val = self.read_discriminant(self.place_to_op(place)?)?.0;
+                let op = self.eval_place_to_op(place, None)?;
+                let discr_val = self.read_discriminant(op)?.0;
```
Comment (RalfJung) [inline src/librustc_mir/interpret/step.rs]: This is a nice catch. :)

File: src/librustc_mir/const_eval.rs [hunk 10]
```diff
@@ -520,25 +503,41 @@ fn validate_and_turn_into_const<'a, 'tcx>(
...
+            }
+            layout::Abi::ScalarPair(..) if mplace.layout.ty.is_slice() => {
+                let (a, b) = ecx.try_read_immediate_from_mplace(mplace)?.unwrap().to_scalar_pair()?;
+                ConstValue::Slice(a, b.to_usize(&ecx)?)
+            },
+            _ => ConstValue::ByRef,
+        };
+        let ptr = Pointer::from(constant.alloc_id);
+        let alloc = constant.alloc;
+        Ok(ty::Const { val, ty: mplace.layout.ty, alloc: Some((alloc, ptr))})
```
Comment (RalfJung) [inline src/librustc_mir/const_eval.rs]: The code above here is basically `mplace_to_const`, right? And we have the same in `const_field`, don't we? Couldn't that become a helper function?

File: src/librustc_mir/const_eval.rs [hunk 11]
```diff
@@ -475,21 +434,45 @@ pub fn const_field<'a, 'tcx>(
...
+                alloc,
+                field_ptr,
+            )),
+        })
     })();
     result.map_err(|error| {
         let err = error_to_const_error(&ecx, error);
+        // FIXME(oli-obk): I believe this is unreachable and we can just ICE here. Since a constant
+        // is checked for validity before being in a place that could pass it to `const_field`,
+        // we can't possibly have errors. All fields have already been checked.
```
Comment (RalfJung) [inline src/librustc_mir/const_eval.rs]: Could/should this FIXME be ported to the other PR?

Comment (oli-obk) [reply src/librustc_mir/const_eval.rs]: I'm trying this out locally in a different branch, so there's no reason to add it to the source tree

**Comment by RalfJung** [comment]
I believe this got superseded by https://github.com/rust-lang/rust/pull/58511 and can be closed?
